validator/accounts/v2/wallet/wallet.go

```diff
@@ -100,35 +100,82 @@ func New(cfg *Config) *Wallet {
 	}
 }
 
-// Exists check if a wallet at the specified directory
-// exists and has valid information in it.
-func Exists(walletDir string) error {
+// Exists checks if directory at walletDir exists
+func Exists(walletDir string) (bool, error) {
 	dirExists, err := fileutil.HasDir(walletDir)
 	if err != nil {
-		return errors.Wrap(err, "could not parse wallet directory")
+		return false, errors.Wrap(err, "could not parse wallet directory")
 	}
 	if dirExists {
-		isEmptyWallet, err := isEmptyWallet(walletDir)
-		if err != nil {
-			return errors.Wrap(err, "could not check if wallet has files")
+		return true, nil
+	}
+	return false, nil
+}
+
+// isValidWallet checks if a folder contains a key directory such as `derived`, `remote` or `direct`.
+// Returns true if one of those subdirectories exist, false otherwise.
+func isValidWallet(walletDir string) (bool, error) {
+	expanded, err := fileutil.ExpandPath(walletDir)
+	if err != nil {
+		return false, err
+	}
+	f, err := os.Open(expanded)
+	if err != nil {
+		return false, err
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			log.Debugf("Could not close directory: %s", expanded)
 		}
-		if isEmptyWallet {
-			return ErrNoWalletFound
+	}()
+	names, err := f.Readdirnames(-1)
+	if err == io.EOF {
+		return false, nil
+	}
+
+	for _, n := range names {
+		// Nil error means input name is `derived`, `remote` or `direct`
+		_, err := v2keymanager.ParseKind(n)
+		if err == nil {
+			return true, nil
 		}
-		return nil
 	}
-	return ErrNoWalletFound
+
+	return false, nil
+}
+
+// ExistsAndValid returns a nil error if walletDir both exists and is valid, and if not, returns the error
+func ExistsAndValid(walletDir string) error {
```

Comment:
I don't think this function is necessary. I would rather make `isValid` public and use `Exists() && IsValid()` everywhere. Note that `&&` short-circuits, meaning that the second operand will not be evaluated if the first operand is `false`.

Reply:
If you wish to keep this function, then it should return `(bool, error)` and return an error only in exceptional scenarios.

Reply:
```
func ExistsAndValid(walletDir string) (bool, error) {
	dirExists, err := Exists(walletDir)
	if err != nil {
		return false, errors.Wrap(err, "could not check if wallet exists")
	}
	return dirExists, nil

	var valid bool
	if valid, err = isValidWallet(walletDir); err != nil {
		return false, errors.Wrap(err, "could not check if wallet is valid")
	}
	return valid, nil
}
```

Reply:
I don't think I can make use of shortcircuit here since both Exists and IsValid return two variables.  For example in OpenWallet (line 210) same file, this is what it is now currently in the PR:
```
	if err := ExistsAndValid(cfg.WalletDir); err != nil {
		return nil, err
	}
```

and this is what it turns into:  

```
	exists, err := Exists(cfg.WalletDir)
	if err != nil {
		return nil, errors.Wrap(err, "could not check if wallet exists")
	}
	if exists {
		valid, err := IsValid(cfg.WalletDir)
		if err != nil{
			return nil, errors.Wrap(err, "could not check if wallet is valid")
		}
		if !valid {
			return nil, errors.New("directory does not contain valid wallet")
		}
	}
```

And there will be repetition of this in other areas as a result as well.  I will make the change and see what it looks like though.

Reply:
Now I feel like the duplication is larger. I think one part of the issue is ErrNoWalletFound -- in the original code this was returned from the combo function (Exists() which tested for existence and validity) if the directory didn't exist **or** if it did but was totally empty.

Reply:
Because ErrNoWalletFound can represent these two different situations, it still crosses the gap between Exists and IsValid and the code that utilizes these functions end up looking for it from each function.